File: 백준/Gold/1062. 가르침/가르침.cc
/*
집합을브루트포스로만든다하면2^26의시간복잡도이고이거는10^6~7이므로1초가안넘기때메가능
*/

#include "bits/stdc++.h"

using namespace std;

int N, K;
vector<string> words;
int ans = 0;
int X;
int Y[50];

int getCnt()
{
    int cnt = 0;
    for (int i = 0; i < N; i++)
    {
        if ((Y[i] & X) == Y[i])
            cnt++;
    }
    return cnt;
}

void dfs(int idx,int k)
{
    if (idx == 26||k==K)
    {
        // for(int i=0;i<26;i++)
        // {
        //     if(X&(1<<i))
        //         cout<<(char)('a'+i)<<" ";
        // }
        // cout<<'\n';
        ans = max(ans, getCnt());
        return;
    }
    if (X & (1 << idx))
        dfs(idx + 1,k);
    else
    {
        X ^= (1 << idx);
        dfs(idx + 1,k+1);
        X ^= (1 << idx);
        dfs(idx + 1,k);
    }
}

int main()
{
    ios_base::sync_with_stdio(0);
    cin.tie(0);
    cin >> N >> K;
    for (int i = 0; i < N; i++)
    {
        string s;
        cin >> s;
        words.push_back(s);
        for (int j = 0; j < s.size(); j++)
            Y[i] |= (1 << (s[j] - 'a'));
    }
    X |= (1 << ('a' - 'a')); // 0
    X |= (1 << ('n' - 'a')); // 13
    X |= (1 << ('t' - 'a')); // 19
    X |= (1 << ('i' - 'a')); // 8
    X |= (1 << ('c' - 'a')); // 2
    if (K < 5)
    {
        cout << 0;
        return 0;
    }
    dfs(0,5);
    cout << ans;
}

File: 백준/Gold/1062. 가르침/가르침.java
import java.io.*;
import java.util.*;

public class Main {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    StringTokenizer st;
    int n, k;
    String[] str = new String[55];
    boolean[] selected = new boolean[33];
    int ans = 0;
    char[] condi = {'a', 'n', 't', 'i', 'c'};

    void dfs(int idx, int start) {
        for (int i = 0; i < condi.length; i++) {
            if (start > condi[i] - 'a' && !selected[condi[i] - 'a'])
                return;
        }
        if (idx >= k) {
            int cnt = 0;
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < str[i].length(); j++) {
                    if (!selected[str[i].charAt(j) - 'a'])
                        break;
                    if (j == str[i].length() - 1)
                        cnt++;
                }
                ans = Math.max(ans, cnt);
            }
            return;
        }
        for (int i = start; i < 26; i++) {
            selected[i] = true;
            dfs(idx + 1, i + 1);
            selected[i] = false;
        }
    }

    void solution() throws Exception {
        st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        k = Integer.parseInt(st.nextToken());
        for (int i = 0; i < n; i++)
            str[i] = br.readLine();
        dfs(0, 0);
        System.out.println(ans);
    }

    public static void main(String[] args) throws Exception {
        new Main().solution();
    }
}